cmd/nvidia-cdi-hook/cudacompat/cudacompat.go

```diff
@@ -0,0 +1,238 @@
+		return nil
+	}
+	if !containerRoot.hasPath("/etc/ld.so.conf.d") {
+		// If the /etc/ld.so.conf.d folder does not exist in the container, the hook is a no-op.
+		return nil
+	}
+
+	libs, err := containerRoot.globFiles(filepath.Join(cudaCompatPath, "libcuda.so.*.*"))
+	if err != nil {
+		m.logger.Warningf("Failed to find CUDA compat library: %w", err)
+		return nil
+	}
+
+	if len(libs) == 0 {
+		return nil
+	}
+
+	if len(libs) != 1 {
+		m.logger.Warningf("Unexpected number of CUDA compat libraries: %v", libs)
+		return nil
+	}
+
+	compatDriverVersion := strings.TrimPrefix(filepath.Base(libs[0]), "libcuda.so.")
+	compatMajor := strings.SplitN(compatDriverVersion, ".", 2)[0]
+
+	hostDriverVersion := cfg.hostDriverVersion
+	driverMajor := strings.SplitN(hostDriverVersion, ".", 2)[0]
+
+	if driverMajor < compatMajor {
+		return m.createConfig(string(containerRoot), []string{cudaCompatPath})
+	}
+	return nil
+}
+
+// A root is used to add basic path functionality to a string.
+type root string
+
+// hasPath checks whether the specified path exists in the root.
+func (r root) hasPath(path string) bool {
+	resolved, err := r.resolve(path)
+	if err != nil {
+		return false
+	}
+	if _, err := os.Stat(resolved); err != nil && os.IsNotExist(err) {
+		return false
+	}
+	return true
+}
+
+// globFiles matches the specified pattern in the root.
+// The files that match must be regular files.
+func (r root) globFiles(pattern string) ([]string, error) {
+	patternPath, err := r.resolve(pattern)
+	if err != nil {
+		return nil, err
+	}
+	matches, err := filepath.Glob(patternPath)
+	if err != nil {
+		return nil, err
+	}
+	var files []string
+	for _, match := range matches {
+		info, err := os.Lstat(match)
+		if err != nil {
+			return nil, err
+		}
+		// Ignore symlinks.
+		if info.Mode()&os.ModeSymlink != 0 {
+			continue
+		}
+		// Ignore directories.
+		if info.IsDir() {
+			continue
+		}
+		files = append(files, match)
+	}
+	return files, nil
+}
+
+// resolve returns the absolute path including root path.
+// Symlinks are resolved, but are guaranteed to resolve in the root.
+func (r root) resolve(path string) (string, error) {
+	absolute := filepath.Clean(filepath.Join(string(r), path))
+	return symlink.FollowSymlinkInScope(absolute, string(r))
+}
+
+// createConfig creates (or updates) /etc/ld.so.conf.d/00-compat-<RANDOM_STRING>.conf in the container
+// to include the required paths.
+// Note that the 00-compat prefix is chosen to ensure that these libraries have
+// a higher precedence than other libraries on the system.
+func (m command) createConfig(root string, folders []string) error {
+	if len(folders) == 0 {
+		m.logger.Debugf("No folders to add to /etc/ld.so.conf")
+		return nil
+	}
+
+	if err := os.MkdirAll(filepath.Join(root, "/etc/ld.so.conf.d"), 0755); err != nil {
+		return fmt.Errorf("failed to create ld.so.conf.d: %w", err)
+	}
+
+	configFile, err := os.CreateTemp(filepath.Join(root, "/etc/ld.so.conf.d"), "00-compat-*.conf")
```

Comment:
Question -- since the file we are creating is not a temporary file, I am assuming `CreateTemp` is used here for generating the random string suffix?

Reply:
Yes we're leveraging that for the random string. I'm sure we could come up with an alternative -- or even just use a timestamp -- if you prefer.

Reply:
Let's stick with the implementation as is.